wallets/wrap.js

```diff
@@ -0,0 +1,166 @@
+          case 16: // basic multi-part payment
+          case 17:
+          case 25: // blinded paths
+          case 48: // TLV payment data
+          case 49:
+          case 149: // trampoline routing
+          case 151: // electrum trampoline routing
+            break
+          default:
+            throw new Error(`Unsupported feature bit: ${f.bit}`)
+        }
+      }
+    } else {
+      throw new Error('Invoice features are missing')
+    }
+
+    // validate the payment hash
+    if (inv.id) {
+      wrapped.id = inv.id
+    } else {
+      throw new Error('Invoice hash is missing')
+    }
+
+    // validate the description
+    if (description && descriptionHash) {
+      throw new Error('Only one of description or descriptionHash is allowed')
+    } else if (description) {
+      // use our wrapped description
+      wrapped.description = description
+    } else if (descriptionHash) {
+      // use our wrapped description hash
+      wrapped.description_hash = descriptionHash
+    } else if (inv.description_hash) {
+      // use the invoice description hash
+      wrapped.description_hash = inv.description_hash
+    } else {
+      // use the invoice description
+      wrapped.description = inv.description
+    }
+
+    // validate the expiration
+    if (new Date(inv.expires_at) < new Date(Date.now() + INCOMING_EXPIRATION_BUFFER_MSECS)) {
+      throw new Error('Invoice expiration is too soon')
+    } else if (new Date(inv.expires_at) > new Date(Date.now() + MAX_EXPIRATION_INCOMING_MSECS)) {
+      // trim the expiration to the maximum allowed with a buffer
+      wrapped.expires_at = new Date(Date.now() + MAX_EXPIRATION_INCOMING_MSECS - INCOMING_EXPIRATION_BUFFER_MSECS)
+    } else {
+      // give the existing expiration a buffer
+      wrapped.expires_at = new Date(new Date(inv.expires_at).getTime() - INCOMING_EXPIRATION_BUFFER_MSECS)
+    }
+
+    // get routing estimates
+    const { routingFeeMsat, timeLockDelay } =
+      await estimateRouteFee({
+        lnd,
+        destination: inv.destination,
+        mtokens: inv.mtokens,
+        request: bolt11,
+        timeout: FEE_ESTIMATE_TIMEOUT_SECS
+      })
+
+    const { current_block_height: blockHeight } = await getHeight({ lnd })
+    /*
+      we want the incoming invoice to have MIN_SETTLEMENT_CLTV_DELTA higher final cltv delta than
+      the expected ctlv_delta of the outgoing invoice's entire route
+
+      timeLockDelay is the absolute height the outgoing route is estimated to expire in the worst case.
+      It excludes the final hop's cltv_delta, so we add it. We subtract the blockheight,
+      then add on how many blocks we want to reserve to settle the incoming payment,
+      assuming the outgoing payment settles at the worst case (ie largest) height.
+    */
+    wrapped.cltv_delta = toPositiveNumber(
+      toPositiveNumber(timeLockDelay) + toPositiveNumber(inv.cltv_delta) -
+      toPositiveNumber(blockHeight) + MIN_SETTLEMENT_CLTV_DELTA)
```

Comment:
I see, this is one of the critical parts. It makes sense to me as follows:

`estimateRouteFee` returns `timeLockDelay` which is the worst case for how many blocks the payment (given destination and amount) may take before the HTLC expires. This means we should wait at least that long. Since it's in absolute block height, we subtract the current block height to get a relative value.

We add the CLTV delta of the final hop to the receiver (`inv.cltv_delta`) since it's not included in `timeLockDelay` (not sure why not though).

We add `MIN_SETTLEMENT_CLTV_DELTA` (= 80) to give ourselves enough time to claim funds on-chain.

There is only one case in which _our_ CLTV delta becomes relevant _for us_: if our peer on the outgoing route decides to settle the payment on-chain; revealing the preimage (HTLC-success tx). We then have 80 blocks to use this preimage to settle the incoming HTLC (on-chain or off-chain) because when this happens, our peer on the incoming route must at least wait 80 blocks per our CLTV delta before they can claim their funds back on-chain (HTLC-timeout tx). This makes sure we always get paid when we forwarded the payment.

In the case our peer on the outgoing route is unresponsive, we have to wait the CLTV delta proposed by that peer (which we accepted) before our HTLC-timeout tx is valid. This makes sure we can always claim back our own funds on-chain. This would be a forced channel close with the unresponsive peer.[^1]

[^1]: It's not clear to me if that force-closure propagates back but I think not since responsive peers can still update their channel with a commitment tx that dropped this HTLC via normal revocation.

_`min_final_cltv_expiry_delta` of invoices as mentioned in BOLT11 was simply renamed to `cltv_delta` by `ln-service`._